In [1]:
import pandas as pd
 
train = pd.read_csv('train_ctrUa4K.csv') 
train.head(100)

Loan_ID  Gender Married Dependents     Education Self_Employed  \
0   LP001002    Male      No          0      Graduate            No   
1   LP001003    Male     Yes          1      Graduate            No   
2   LP001005    Male     Yes          0      Graduate           Yes   
3   LP001006    Male     Yes          0  Not Graduate            No   
4   LP001008    Male      No          0      Graduate            No   
..       ...     ...     ...        ...           ...           ...   
95  LP001326    Male      No          0      Graduate           NaN   
96  LP001327  Female     Yes          0      Graduate            No   
97  LP001333    Male     Yes          0      Graduate            No   
98  LP001334    Male     Yes          0  Not Graduate            No   
99  LP001343    Male     Yes          0      Graduate            No   

    ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0              5849                0.0         NaN             360.0   
1              4583             1508.0       128.0             360.0   
2              3000                0.0        66.0             360.0   
3              2583             2358.0       120.0             360.0   
4              6000                0.0       141.0             360.0   
..              ...                ...         ...               ...   
95             6782                0.0         NaN             360.0   
96             2484             2302.0       137.0             360.0   
97             1977              997.0        50.0             360.0   
98             4188                0.0       115.0             180.0   
99             1759             3541.0       131.0             360.0   

    Credit_History Property_Area Loan_Status  
0              1.0         Urban           Y  
1              1.0         Rural           N  
2              1.0         Urban           Y  
3              1.0         Urban           Y  
4              1.0         Urban           Y  
..             ...           ...         ...  
95             NaN         Urban           N  
96             1.0     Semiurban           Y  
97             1.0     Semiurban           Y  
98             1.0     Semiurban           Y  
99             1.0     Semiurban           Y  

[100 rows x 13 columns]

In [2]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 614 entries, 0 to 613
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Loan_ID            614 non-null    object 
 1   Gender             601 non-null    object 
 2   Married            611 non-null    object 
 3   Dependents         599 non-null    object 
 4   Education          614 non-null    object 
 5   Self_Employed      582 non-null    object 
 6   ApplicantIncome    614 non-null    int64  
 7   CoapplicantIncome  614 non-null    float64
 8   LoanAmount         592 non-null    float64
 9   Loan_Amount_Term   600 non-null    float64
 10  Credit_History     564 non-null    float64
 11  Property_Area      614 non-null    object 
 12  Loan_Status        614 non-null    object 
dtypes: float64(4), int64(1), object(8)
memory usage: 62.5+ KB


In [3]:
train.isnull().sum()

Loan_ID               0
Gender               13
Married               3
Dependents           15
Education             0
Self_Employed        32
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount           22
Loan_Amount_Term     14
Credit_History       50
Property_Area         0
Loan_Status           0
dtype: int64

So, there are missing values on many variables including the Gender, Married, LoanAmount variable. Next, we will remove all the rows which contain any missing values in them:

In [4]:
train = train.dropna()
train.isnull().sum()

Loan_ID              0
Gender               0
Married              0
Dependents           0
Education            0
Self_Employed        0
ApplicantIncome      0
CoapplicantIncome    0
LoanAmount           0
Loan_Amount_Term     0
Credit_History       0
Property_Area        0
Loan_Status          0
dtype: int64

Now there are no missing values in the dataset. Next, we will separate the dependent (Loan_Status) and the independent variables:

In [5]:

X = train[['Gender', 'Married', 'ApplicantIncome', 'LoanAmount', 'Credit_History']]
y = train.Loan_Status
X.shape, y.shape

((480, 5), (480,))

In [6]:
#X=pd.get_dummies(X,columns=['Gender','Married'])
import category_encoders as ce
encoder= ce.OrdinalEncoder(cols=['Gender'],return_df=True,
                           mapping=[{'col':'Gender',
'mapping':{'Male':0,'Female':1}}])

X_transformed = encoder.fit_transform(X)



/root/AutomatedLoanPredict/loan/lib/python3.8/site-packages/category_encoders/utils.py:21: FutureWarning: is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead
  elif pd.api.types.is_categorical(cols):


In [7]:
encoder1= ce.OrdinalEncoder(cols=['Married'],return_df=True,
                           mapping=[{'col':'Married',
'mapping':{'No':0,'Yes':1}}])

X = encoder1.fit_transform(X_transformed)

In [8]:
X.head()

Gender  Married  ApplicantIncome  LoanAmount  Credit_History
1       0        1             4583       128.0             1.0
2       0        1             3000        66.0             1.0
3       0        1             2583       120.0             1.0
4       0        0             6000       141.0             1.0
5       0        1             5417       267.0             1.0

In [9]:
X.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 480 entries, 1 to 613
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Gender           480 non-null    int64  
 1   Married          480 non-null    int64  
 2   ApplicantIncome  480 non-null    int64  
 3   LoanAmount       480 non-null    float64
 4   Credit_History   480 non-null    float64
dtypes: float64(2), int64(3)
memory usage: 22.5 KB


For this particular project, I have only picked 5 variables that I think are most relevant. These are the Gender, Marital Status, ApplicantIncome, LoanAmount, and Credit_History and stored them in variable X. Target variable is stored in another variable y. And there are 480 observations available. Next, let’s move on to the model building stage.

Here, we will first split our dataset into a training and validation set, so that we can train the model on the training set and evaluate its performance on the validation set.

In [10]:
from sklearn.model_selection import train_test_split
x_train, x_cv, y_train, y_cv = train_test_split(X,y, test_size = 0.2, random_state = 10)


We have split the data using the train_test_split function from the sklearn library keeping the test_size as 0.2 which means 20 percent of the total dataset will be kept aside for the validation set. Next, we will train the random forest model using the training set:

In [11]:
from sklearn.ensemble import RandomForestClassifier 
model = RandomForestClassifier(max_depth=5, random_state = 15) 
model.fit(x_train, y_train)

RandomForestClassifier(max_depth=5, random_state=15)

Here, I have kept the max_depth as 4 for each of the trees of our random forest and stored the trained model in a variable named model. Now, our model is trained, let’s check its performance on both the training and validation set:

In [12]:
from sklearn.metrics import accuracy_score
pred_cv = model.predict(x_cv)
accuracy_score(y_cv,pred_cv)

0.8020833333333334

The model is 80% accurate on the validation set. Let’s check the performance on the training set too:

In [13]:
pred_train = model.predict(x_train)
accuracy_score(y_train,pred_train)

0.8333333333333334

In [14]:
# saving the model 
import pickle 
pickle_out = open("predictor.pkl", mode = "wb") 
pickle.dump(model, pickle_out) 
pickle_out.close()